### ISSUE TITLE
BattleScribe rounding inconsequence on iOS

### ISSUE THREAD

**Issue by BSDataAnon** [BSData/wh40k-7th-edition#1775]
The issue is a problem with BattleScribe rounding: see @Jonskichov [comment in this issue](https://github.com/BSData/wh40k/issues/1775#issuecomment-178088573).

----
### OP:
**File:** Chaos Space Marines - Codex.catz

**BattleScribe version:** 1.15.09

**Platform:** iPhone / iPod / iPad

**Dropbox:** Yes

**Description:** Error message when selecting a Legacy of Ruin on a tank:  Maximum 0 is allowed.

Page 187 of IA13, any vehicule in a Codex: Chaos Space Marine army and any of its supplement codexes are allowed to take 1 legacy of ruin per 1000 points.

**Comment by cartag** [comment]
Can you give some more info? If your roster is set to be 0 points max, it will give you an error. If your vehicle has daemonic possession, it will give you an error. Can you please explain how the roster is set up so we can duplicate the issue?

**Comment by Drachnyen** [comment]
Hello,

I just reproduced it,  steps to reproduce:
1. Create new roster for 1500 points
2. select combined arms detachment for chaos space marine codex (v186)
3. Add stuff to reach 1500
4. Select any regular tank (like vindicator or predator)
5. Assigning any ruin brings error, max ruin = 0

**Comment by cartag** [comment]
I just tried that same thing on both my iPhone and my computer. I can't replicate on either. Can you attach a screenshot and/or the actual roster file here so I can take a look at it?

**Comment by cartag** [comment]
Start new roster.
Set limit to 1500 pts
Show FW
Add Vindicator
Add Ruin
No error.
Add second Vindicator
Add Ruin
Get error, correctly.

**Comment by penguin20000** [comment]
@cartag not sure if it helps but i get the same error. Not had chance to look at the data files though.

The only way i can get it to happen is with the points limit set to < 1000 for the roster though

![screen](https://cloud.githubusercontent.com/assets/10774422/12394068/8c4c270a-bdf1-11e5-8bea-44c7645cdcfd.PNG)

**Comment by Drachnyen** [comment]
Hi @cartag,

Retried on my Ipad using your recipe, I get the same error as @penguin20000 

I have included the test roster

[Test 3.zip](https://github.com/BSData/wh40k/files/94562/Test.3.zip)

Thanks,

**Comment by hisop** [comment]
@penguin20000  see how you have 150/0pts? the roster checks how many points your list is set to, as it only allows 1 ruin per full 1000 points so in 750 its 0 in 1500 its 1 in 2000 its 2 ect. your list above is set to 0 this the list correctly tells you you cant have any.

@Drachnyen  just loaded your roster and its fine for me this could be IOS specific, could i get the version of the catalog and the version of battle scribe your using. will test on ios tonight.

**Comment by Drachnyen** [comment]
@hisop 

IOS Battlescribe: 1.15.09
Game System: Warhammer 40,000 7th edition Edition V44
Catalogue: Chaos Space Marines: Codex (2012) v186

![file_000](https://cloud.githubusercontent.com/assets/16340056/12395093/b33d277a-bdcd-11e5-86d8-890d5b8d626d.png)

**Comment by hisop** [comment]
Will look at IOS as soon as i get home, android and windows are clear.

**Comment by Drachnyen** [comment]
Thanks! :)

**Comment by penguin20000** [comment]
@hisop Ahh that'll be it. I apologise ignore my ramblings :)

**Comment by hisop** [comment]
@Drachnyen @penguin20000  Haha, its no problem :)

**Comment by cartag** [comment]
Well, we've found a bug in the iOS version of Battlescribe now. It's set to allow 1 per every 1000 points in the points limit. For some reason this isn't registering properly in iOS version. 

I've found a similar issue already where it's not "show"ing properly by criteria, it doesn't hide by default and then show when the criteria is met. Instead I have to set it to always show, but hide if this criteria is not present. 

Tl;dr: I'm trying to find a workaround.

**Comment by cartag** [comment]
Ok, so I think I have it narrowed down thanks to a eureka moment after coffee.

On a whim I changed the roster limit to 2000, checked again, no error. Add a second one, get an error that max is 1. 

In the catalog, it says it increments maxInForce (MIF) by 1 for each 1000 pts, starting at 0. MaxInRoster (MIR) starts at -1. At 1000 points, incrementing MIF would take it to 1, but MIR would go to 0. At 2000 pts, MIF would be 2, MIR 1. 

If you could all go and throw some votes at this it'll flag it on the feedback forum. http://battlescribe.uservoice.com/forums/55066-general/suggestions/11491731-ios-not-incrementing-properly

**Comment by hisop** [comment]
Voted,i hate battlescribe bugs as it will sit here for an age.

**Comment by amis92** [comment]
@hisop now it's quite probably going to change. ;)
Check out: https://github.com/BSData/bsdata/issues/92

**Comment by amis92** [comment]
Also, is this to be closed?

**Comment by hisop** [comment]
Nice :) Should i test on the alpha? (this bug i mean)

**Comment by amis92** [comment]
Well, it's a long way to be released, but why not - will give you an idea if this is already fixed or not so much :)

**Comment by hisop** [comment]
Well i will wait until create roster is working properly has a few issues :P so until then i propose this remains (to remind me)

**Comment by cartag** [comment]
Might want to wait until I test it out to see if we need to update the data files as well ;) I was being lazy, I just woke up. Checking on it now.

**Comment by cartag** [comment]
From what I could manage to do in the CSM catalog, there are still issues with point total limits. I vote to leave it open for now until the next round of Alpha bug fixes go through.

**Comment by Jonskichov** [comment]
I think I know what this is...

Rounding of decimals to integers in Java is handled differently than C# (used on iOS).

I'll make sure the rounding is explicit in future releases.

As this should be a very rare occurrence, I'll roll this into the next major update (rather than a small update to fix just this).

**Comment by amis92** [comment]
Can anyone confirm if this is still an issue in Alphas? If it's not, please close it.

**Comment by amis92** [comment]
@Jonskichov can you confirm you've fixed this?

**Comment by Drachnyen** [comment]
Not fixed. 

I am still experiencing this issue with BS2 on IOS

**Comment by francoisbeaulieu** [comment]
Is this fixed in BS 2.00.07 for iOS?

**Comment by cartag** [comment]
Seems resolved.
